### ISSUE TITLE
Replace decompositions in torchinductor/decompositions with refs/decomps in pytorch proper

### ISSUE THREAD

**Issue by ngimel** [pytorch/torchdynamo#801]
We should audit existing decompositions and replace them with refs/decomps where appropriate, e.g. `clamp` decomposition currently incorrectly type promotes and doesn't handle non-number boundaries (see also #594).

**Comment by Chillee** [comment]
It's not just about decompositions, it's also about lowerings I believe.

**Comment by ngimel** [comment]
Lowerings I'm not so sure about (yeah we need to fix type promotion on lowerings, but I'm not sure refs + prims are going to be helpful directly, rather, we'd use type promotion utilities from primtorch.

**Comment by jansel** [comment]
I'd start with just the decomps.  Much of the stuff we do in lowering isn't expressible in primtorch.

**Comment by ezyang** [comment]
@jansel Is that still true even with arange style decomps?

**Comment by jansel** [comment]
Yeah, the arange style decomps will generate worse code than lowerings because [index simplification](https://github.com/pytorch/torchdynamo/blob/f6bede60718699dc77ce7c7a1b3458c0e7881876/torchinductor/sizevars.py#L63) and [tiling](https://github.com/pytorch/torchdynamo/blob/f6bede60718699dc77ce7c7a1b3458c0e7881876/torchinductor/codegen/triton.py#L1150) won't work on it.  It is still the best alternative for writing a decomp though.

There is also no way to express masked loads as a decomp.

**Comment by Chillee** [comment]
For lowerings I was referring to the trivial ones (I.e. pointwise ops), where inductor duplicates type promotion logic.

**Comment by Chillee** [comment]
@jansel It’s plausible to me that we can make arange decomps work with a little bit of work to statically analyze arange.

**Comment by jansel** [comment]
I agree it is plausible to write static analysis convert the `ops.indexing_expr()` + arbitrary math + `ops.indirect_indexing()` back into a sympy pure expression without crossing back and forth between the indexing and values regimes.

Writing that analysis doesn't seem like a very productive use of time at the moment though.  We already have a solution for this that works well, and lots of other tasks that seem higher priority to me.

**Comment by ezyang** [comment]
The long term argument for this analysis is that decomps represent user programmable frontend UX (so if an arange optimizes well, that means users can use it to drive inductor), but lowerings are not (unless you plan to make lowerings public API, but that's probably a bad idea if we ever want to refactor this part of the codebase.)

**Comment by ezyang** [comment]
I agree short term getting type promotion working correctly via decomps is probably the best bang for buck

**Comment by jansel** [comment]
Longer term I think we should build that optimization.  It just seems lower priority for now.

The current arange will perform decently well, so I think it is fine for users to write it today.  It will miss things like merging contiguous dimensions and tiling, but for lots of kernels those optimizations are not needed for good performance.

**Comment by Chillee** [comment]
There's actually another reason we might want to use decomps instead of lowerings. Using decomps allows us to leverage autograd by decomposing *before* autograd, while using lowerings requires us to also implement a lowering for, say, `upsample_2d_backward`.

**Comment by ngimel** [comment]
That's a valid reason, but we should check perf of upsampling decomp vs lowering. Also, `upsample_2d_backward` may not be the worst thing to have lowering for, but it closes the door on higher order grads

**Comment by Chillee** [comment]
From my benchmarking, it seems like the decomposition for `aten.upsample_bilinear2d.vec` actually performs on par/better than the lowering now: https://github.com/pytorch/torchdynamo/pull/934
